### PR TITLE
Add new case for positivity checker: type cannot occur as arg of bound var

### LINKED ISSUE #2540: Unsound positive checker when using the (new) typechecker (--new-typechecker)
As in: https://github.com/anoma/juvix/pull/2506#issuecomment-1811659513

The positive checker with the type checker using `--new-typechecker` allows non-strictly positive definitions.

```
module NegativeEvil;

type Evil (f : Type -> Type) :=
  | magic : f (Evil f) -> Evil f;
```

First, the `Evil` data type is not strictly positive but it passes our checker, so we have an issue. Second, ignoring the positivity condition over the Evil definition for a moment, consider the rest of this program.  

```
type Empty := mkEmpty {}; 

toEmpty (A : Type) : Type := A -> Empty;

panic : toEmpty (Evil toEmpty) 
  | x@(magic h) := h x ;

god : Empty := panic (magic panic); 
```

The above produces the following error, 
```
The expression panic has type:
  Evil λ! {A := A → Empty} → Empty
but is expected to have type:
  _ (Evil _)
```

The message is a bit difficult to understand, at least to me, so we must fix it too.
Also, this should typecheck ignoring the positivity condition.

Similarly as with `Evil`, one could derive a term of `Empty` using for example, other types like:

```
type Free (f : Type -> Type) (A : Type)
 : Type :=
| Leaf : A -> Free f A
| Branch : f (Free f A) -> Free f A;
```
We can include these examples in a future PR fixing the pos. checker.

### PR DESCRIPTION
**Pull request by jonaprieto**
This PR fixes our positivity checker to support inductive definitions with type families as type parameters. This kind of ind. type is type-checked using the global flag `--new-type checker.`

For example, the following definition is not allowed:

```
module Evil;

type Evil (f : Type -> Type) :=
  magic : f (Evil f) -> Evil f;
```

- Closes #2540

### CODE REVIEW

File: src/Juvix/Compiler/Internal/Translation/FromInternal/Analysis/Positivity/Checker.hs [hunk 1]
```diff
@@ -28,157 +22,202 @@ type CheckPositivityEffects r =
...
+                      _checkPositivityArgsConstructorName =
+                        ctor ^. inductiveConstructorName,
+                      _checkPositivityArgsInductiveName = ty ^. inductiveName,
+                      _checkPositivityArgsRecursionLimit = numInductives,
+                      _checkPositivityArgsErrorReference = Nothing,
+                      _checkPositivityArgsTypeOfConstructor = typeOfConstr
+                    }
+                )
+          )
+          (constructorArgs (ctor ^. inductiveConstructorType))
```
Comment (janmasrovira) [inline src/Juvix/Compiler/Internal/Translation/FromInternal/Analysis/Positivity/Checker.hs:60]: When the mapped function takes multiple lines I usually prefer to use `forM_` inetead.
```suggestion
        forM_
          (constructorArgs (ctor ^. inductiveConstructorType))
          $ \typeOfConstr ->
              checkStrictlyPositiveOccurrences
                ( CheckPositivityArgs
                    { _checkPositivityArgsInductive = ty,
                      _checkPositivityArgsConstructorName =
                        ctor ^. inductiveConstructorName,
                      _checkPositivityArgsInductiveName = ty ^. inductiveName,
                      _checkPositivityArgsRecursionLimit = numInductives,
                      _checkPositivityArgsErrorReference = Nothing,
                      _checkPositivityArgsTypeOfConstructor = typeOfConstr
                    }
                )
```

File: src/Juvix/Compiler/Internal/Translation/FromInternal/Analysis/Positivity/Checker.hs [hunk 2]
```diff
@@ -28,157 +22,202 @@ type CheckPositivityEffects r =
...
             | name' `elem` ty ^.. inductiveParameters . each . inductiveParamName -> modify (HashSet.insert name')
           _ -> return ()
 
-        helperApp :: Application -> Sem r ()
-        helperApp tyApp = do
+        goApp :: Application -> Sem r ()
+        goApp tyApp = do
           let (hdExpr, args) = unfoldApplication tyApp
           case hdExpr of
+            ax@(ExpressionIden (IdenAxiom _)) -> do
```
Comment (janmasrovira) [inline src/Juvix/Compiler/Internal/Translation/FromInternal/Analysis/Positivity/Checker.hs:150]:
```suggestion
            ax@ExpressionIden IdenAxiom {} -> do
```

File: src/Juvix/Compiler/Internal/Translation/FromInternal/Analysis/Positivity/Checker.hs [hunk 3]
```diff
@@ -28,157 +22,202 @@ type CheckPositivityEffects r =
...
+              when (isJust $ find (varOrInductiveInExpression name) args) $
+                throwTypeAsArgumentOfBoundVarError ax
+            var@(ExpressionIden (IdenVar _)) -> do
```
Comment (janmasrovira) [inline src/Juvix/Compiler/Internal/Translation/FromInternal/Analysis/Positivity/Checker.hs:153]:
```suggestion
            var@ExpressionIden IdenVar {} -> do
```

File: test/Typecheck/Negative.hs [hunk 4]
```diff
@@ -263,38 +264,46 @@ negPositivityTests :: [NegTest]
 negPositivityTests =
   [ negTest "E1" $(mkRelDir "Internal/Positivity") $(mkRelFile "E1.juvix") $
       \case
-        ErrNoPositivity {} -> Nothing
+        ErrNonStrictlyPositive (ErrTypeInNegativePosition {}) -> Nothing
```
Comment (janmasrovira) [inline test/Typecheck/Negative.hs:267]:
```suggestion
        ErrNonStrictlyPositive ErrTypeInNegativePosition {} -> Nothing
```
Parentheses can also be removed for the rest of the cases

File: test/Typecheck/NegativeNew.hs [hunk 5]
```diff
@@ -146,5 +147,17 @@ arityTests =
       $(mkRelFile "DefaultArgCycleArity.juvix")
       $ \case
         ErrDefaultArgLoop {} -> Nothing
+        _ -> wrongError,
+    negTest "Evil: issue 2540" $(mkRelDir "Internal/Positivity") $(mkRelFile "Evil.juvix") $
+      \case
+        ErrNonStrictlyPositive (ErrTypeAsArgumentOfBoundVar {}) -> Nothing
```
Comment (janmasrovira) [inline test/Typecheck/NegativeNew.hs:153]:
```suggestion
        ErrNonStrictlyPositive ErrTypeAsArgumentOfBoundVar {} -> Nothing
```

File: src/Juvix/Compiler/Internal/Translation/FromInternal/Analysis/Positivity/Checker.hs (hunk 2 above)
Comment (jonaprieto) [reply src/Juvix/Compiler/Internal/Translation/FromInternal/Analysis/Positivity/Checker.hs]: This creates a parsing error.

Comment (jonaprieto) [reply src/Juvix/Compiler/Internal/Translation/FromInternal/Analysis/Positivity/Checker.hs]: The outer parenthesis has to stay.

**Review by janmasrovira — APPROVED** [review]

**Pull request merged**